File: leetcode/0206-reverse-linked-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        RUNTIME: 39 ms (88.78%)
        MEMORY: 15.4 MB (94.38%)
        """
        reversed_list = None
        while head:
            tmp = head.next
            head.next = reversed_list
            reversed_list = head
            head = tmp
        return reversed_list

File: leetcode/0206-reverse-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    /**
     * RUNTIME: 3 ms (98.73%)
     * MEMORY: 8.3 MB (80.16%)
    */
    ListNode* reverseList(ListNode* head) {
        ListNode *tmp, *reversed_list = NULL;
        while (head) {
            tmp = head->next;
            head->next = reversed_list;
            reversed_list = head;
            head = tmp;
        }
        return reversed_list;
    }
};
